### PR TITLE
fix(react-grid): correct calculating start index of loading row for Infinite Scrolling

### LINKED ISSUE #2139: Infinite scrolling sometimes won't render rows
- [x] I have searched this repository's [issues](https://github.com/devexpress/devextreme-reactive/issues) and believe that this is not a duplicate.

### I'm using ...

- [x] React Grid
- [ ] React Chart
- [ ] React Scheduler

### Current Behaviour

Sometimes when scrolling fast with Infinite Scrolling enabled the table won't render rows, as shown in the two images below taken from your own documentation.

image

image

There are some other issues with this problem but they have both been closed.
https://github.com/DevExpress/devextreme-reactive/issues/1253
https://github.com/DevExpress/devextreme-reactive/issues/2062

### PR DESCRIPTION
**Pull request by LazyLahtak**
Fixes #2139

### CODE REVIEW

**Comment by LazyLahtak** [comment]
FPR

File: packages/dx-grid-core/src/plugins/virtual-table-state/helpers.ts [hunk 1]
```diff
@@ -43,12 +43,18 @@ export const mergeRows: MergeRowsFn = (
...
-  virtualRows, newRange, pageSize,
+  virtualRows, newRange, pageSize, referenceIndex,
 ) => {
   const loadedInterval = intervalUtil.getRowsInterval(virtualRows);
   const isAdjacentPage = Math.abs(loadedInterval.start - newRange.start) < 2 * pageSize;
   if (isAdjacentPage) {
-    return intervalUtil.difference(newRange, loadedInterval);
+    const calculatedRange = intervalUtil.difference(newRange, loadedInterval);
+    const firstHalfOfPage = calculatedRange.start - referenceIndex > pageSize / 2;
+    if (calculatedRange.start > referenceIndex && firstHalfOfPage) {
```
Comment (MaximKudriavtsev) [inline packages/dx-grid-core/src/plugins/virtual-table-state/helpers.ts]: Can we use only the next `calculatedRange.start - referenceIndex > pageSize / 2` codition?

if the above condition is true and `pageSize` > 0 this condition `calculatedRange.start > referenceIndex` is true

**Comment by LazyLahtak** [comment]
FPR

**Comment by LazyLahtak** [comment]
FPR

File: packages/dx-grid-core/src/plugins/virtual-table-state/helpers.ts [hunk 2]
```diff
@@ -143,7 +143,7 @@ export const needFetchMorePages: PureComputed<[VirtualRows, number, number], boo
   const { start, end } = intervalUtil.getRowsInterval(virtualRows);
   const loadCount = end - start;
   const topTriggerIndex = start > 0 ? start + pageSize! : 0;
-  const bottomTriggerIndex = end - pageSize!;
+  const bottomTriggerIndex = Math.max(topTriggerIndex + pageSize, end - pageSize! * 1.5);
```
Comment (MaximKudriavtsev) [inline packages/dx-grid-core/src/plugins/virtual-table-state/helpers.ts]: What is a `1.5` value?

Comment (LazyLahtak) [reply packages/dx-grid-core/src/plugins/virtual-table-state/helpers.ts]: Half of page.
Sometimes difference between `bottomTriggerIndex` and `topTriggerIndex` more than one `pageSize` and we can get index, than no trigger update. This is index in interval between `end - pageSize!` and `end - pageSize! * 1.5`
For example, in master try after load press `End` button twice (lazy loading demo)
You will stuck at string 150 as last with `bottomTriggerIndex` === 150 and `index ~ 140`

Comment (LazyLahtak) [reply packages/dx-grid-core/src/plugins/virtual-table-state/helpers.ts]: Sometimes we may catch this bug not in beginning of table. It is quite difficult to catch, but can.

File: packages/dx-grid-core/src/plugins/virtual-table-state/helpers.test.ts [hunk 3]
```diff
@@ -228,10 +229,45 @@ describe('VirtualTableState helpers', () => {
...
+
+      // tslint:disable-next-line: max-line-length
+      it('should caclulate correct if reference index more than the start of a new interval and less than half of page', () => {
+        const loadedInterval = createInterval(100, 400);
+        const newInterval = createInterval(200, 500);
+        const virtualRows = createVirtualRows(loadedInterval);
+        const referenceIndex = 320;
+
+        expect(calculateRequestedRange(virtualRows, newInterval, pageSize, referenceIndex))
+          .toEqual({ start: 300, end: 400 });
```
Comment (ushkal) [inline packages/dx-grid-core/src/plugins/virtual-table-state/helpers.test.ts]: Why do we want rows 300-400 if they're loaded already?

File: packages/dx-grid-core/src/plugins/virtual-table-state/helpers.ts [hunk 4]
```diff
@@ -43,12 +43,17 @@ export const mergeRows: MergeRowsFn = (
...
+  virtualRows, newRange, pageSize, referenceIndex,
 ) => {
   const loadedInterval = intervalUtil.getRowsInterval(virtualRows);
   const isAdjacentPage = Math.abs(loadedInterval.start - newRange.start) < 2 * pageSize;
-  if (isAdjacentPage) {
-    return intervalUtil.difference(newRange, loadedInterval);
+  const calculatedRange = intervalUtil.difference(newRange, loadedInterval);
+  if (isAdjacentPage && calculatedRange !== intervalUtil.empty) {
+    if (calculatedRange.start - referenceIndex > pageSize / 2) {
+      calculatedRange.start -= pageSize;
```
Comment (ushkal) [inline packages/dx-grid-core/src/plugins/virtual-table-state/helpers.ts]: In non-infinite scrolling mode request range shouldn't be affected

File: packages/dx-grid-core/src/plugins/virtual-table-state/helpers.test.ts (hunk 3 above)
Comment (LazyLahtak) [reply packages/dx-grid-core/src/plugins/virtual-table-state/helpers.test.ts]: When we try load next page (rows 400-500) width index less than page 350 (for `pageSize === 100`), we catch index bug. It loads nextPage instead curren when we fast scroll up.
We see at string 320, cache know about previous (200-300) and next page (400-500), and we load 400-500. Then, we get rows combination:
```
[200,
201,
...
300,
400,
401,
...
500,
401,
...
500]
```
For index less than `start - pageSize/2` we need to shift loading interval.

Comment (ushkal) [reply packages/dx-grid-core/src/plugins/virtual-table-state/helpers.test.ts]: Does it reproduce in non-infinite mode?

File: packages/dx-grid-core/src/plugins/virtual-table-state/helpers.ts (hunk 4 above)
Comment (LazyLahtak) [reply packages/dx-grid-core/src/plugins/virtual-table-state/helpers.ts]: You're right, I'll fix it.

**Comment by LazyLahtak** [comment]
FPR

**Comment by LazyLahtak** [comment]
FPR

**Comment by LazyLahtak** [comment]
FPR

**Pull request merged**